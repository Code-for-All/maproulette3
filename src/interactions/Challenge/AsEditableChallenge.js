import _isFinite from 'lodash/isFinite'
import _get from 'lodash/get'

/**
 * AsEditableChallenge adds functionality to a Challenge related to editing.
 */
export class AsEditableChallenge {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  isSourceReadOnly() {
    return _isFinite(this.id) && _get(this, 'actions.total') !== 0
  }

  clearSources() {
    delete this.localGeoJSON
    delete this.overpassQL
    delete this.remoteGeoJson
  }
}

export default challenge => new AsEditableChallenge(challenge)
