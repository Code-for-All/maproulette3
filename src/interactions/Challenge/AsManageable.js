import _isEmpty from 'lodash/isEmpty'

/**
 * AsManageable adds functionality to a Challenge related to management.
 */
export class AsManageable {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  isRebuildable() {
    return !_isEmpty(this.overpassQL)
  }
}

export default challenge => new AsManageable(challenge)
