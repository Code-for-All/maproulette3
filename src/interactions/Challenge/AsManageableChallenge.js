import _isEmpty from 'lodash/isEmpty'
import _isObject from 'lodash/isObject'
import _isFinite from 'lodash/isFinite'
import _get from 'lodash/get'
import _filter from 'lodash/filter'
import _maxBy from 'lodash/maxBy'
import parse from 'date-fns/parse'
import { isCompletionStatus }
       from '../../services/Task/TaskStatus/TaskStatus'

/**
 * AsManageableChallenge adds functionality to a Challenge related to
 * management.
 */
export class AsManageableChallenge {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  isRebuildable() {
    return !_isEmpty(this.overpassQL)
  }

  isComplete() {
    return _get(this, 'actions.total', 0) > 0 &&
           _get(this, 'actions.available') === 0
  }

  completionPercentage() {
    if (_get(this, 'actions.total', 0) > 0) {
      return percentage(this.actions.total,
                        this.actions.total - this.actions.available)
    }
    else {
      return 0
    }
  }

  actionPercentage(action) {
    return percentage(_get(this, 'actions.total'),
                      _get(this, `actions.${action}`, 0))
  }

  completionActivity() {
    return _filter(
      this.activity,
      entry => entry.count > 0 && isCompletionStatus(entry.status)
    )
  }

  mostRecentCompletionActivityEntry() {
    if (_isEmpty(this.activity)) {
      return null
    }

    return _maxBy(this.completionActivity(),
                  entry => parse(entry.date).getTime())
  }

  mostRecentCompletionActivityDate() {
    const entry = this.mostRecentCompletionActivityEntry()
    return _isObject(entry) ? parse(entry.date) : null
  }
}

/**
 * @private
 */
export const percentage = function(total, value, decimalPlaces=1) {
  let percentage = 0
  if (_isFinite(total) && total > 0) {
    percentage = value / total * 100.0
  }

  return Math.round(percentage)
}

export default challenge => new AsManageableChallenge(challenge)
