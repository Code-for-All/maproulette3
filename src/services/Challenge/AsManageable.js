import _isEmpty from 'lodash/isEmpty'

/**
 * AsManageable adds functionality to a Challenge related to management.
 */
export class AsManageable {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  isRebuildable() {
    return !_isEmpty(this.overpassQL) || !_isEmpty(this.remoteGeoJson)
  }
}

export default challenge => new AsManageable(challenge)
