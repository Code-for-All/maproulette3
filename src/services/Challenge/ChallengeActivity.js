import _map from 'lodash/map'
import _groupBy from 'lodash/groupBy'
import _each from 'lodash/each'
import _pick from 'lodash/pick'
import _reduce from 'lodash/reduce'
import { statusMachineName } from '../Task/TaskStatus/TaskStatus'

/**
 * Converts detailed activity (such as from the `status/activity` endpoint) to the
 * traditional summary format provided by other activity endpoints. Normalized
 * activity is returned by challenge id.
 */
export const normalizeDetailedActivity = function(detailedActivity) {
  // Begin by adding the additional fields expected for summary entries
  const normalizedActivity = _map(detailedActivity, entry => {
    entry.status = entry.newStatus
    entry.statusName = statusMachineName(entry.newStatus)
    entry.date = midnightOnDate(entry.created)
    entry.count = 1

    return entry
  })

  // Now we need to group activity by challenge, and then -- for each challenge --
  // accumulate all similar status changes on a single date into a single entry
  // for that status with a count.
  //
  // Start by running through the entries for each challenge
  const activityByChallenge = _groupBy(normalizedActivity, 'challengeId')
  _each(activityByChallenge, (activityWithChallenge, challengeId) => {
    const challengeActivityResults = []

    // Run through the entries for each day for this challenge
    const challengeActivityByDate = _groupBy(activityWithChallenge, 'date')
    _each(challengeActivityByDate, (activityWithDate, date) => {
      // Run through the entries for each status for this day
      const datedChallengeActivityByStatus = _groupBy(activityWithDate, 'status')
      _each(datedChallengeActivityByStatus, (activityWithStatus, status) => {
        // Consolidate all the entries for this day and status into a single
        // entry with a count.
        const consolidatedStatusEntry =
          _reduce(activityWithStatus, (consolidated, entry) => {
            consolidated.count += entry.count
            return consolidated
          })

        // Add this consolidated entry to our challenge activity results, picking
        // just the expected fields.
        challengeActivityResults.push(
          _pick(consolidatedStatusEntry, ['date', 'status', 'statusName', 'count'])
        )
      })
    })

    activityByChallenge[challengeId] = challengeActivityResults
  })

  return activityByChallenge
}

/**
 * @private
 */
export const midnightOnDate = function(isoDateString) {
  return isoDateString.replace(/T.+$/, 'T00:00:00')
}
