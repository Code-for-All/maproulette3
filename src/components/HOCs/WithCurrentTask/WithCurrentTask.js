import React, { Component } from 'react'
import { connect } from 'react-redux'
import { denormalize } from 'normalizr'
import _get from 'lodash/get'
import _omit from 'lodash/omit'
import _isFinite from 'lodash/isFinite'
import _isString from 'lodash/isString'
import _isObject from 'lodash/isObject'
import _debounce from 'lodash/debounce'
import { taskDenormalizationSchema,
         loadCompleteTask,
         loadRandomTaskFromChallenge,
         loadRandomTaskFromVirtualChallenge,
         addTaskComment,
         completeTask } from '../../../services/Task/Task'
import { TaskLoadMethod }
       from '../../../services/Task/TaskLoadMethod/TaskLoadMethod'
import { contactOSMUserURL } from '../../../services/OSMUser/OSMUser'
import { fetchChallengeActions } from '../../../services/Challenge/Challenge'
import { renewVirtualChallenge }
       from '../../../services/VirtualChallenge/VirtualChallenge'

const FRESHNESS_THRESHOLD = 5000 // 5 seconds

/**
 * WithCurrentTask passes down the denormalized task specified in either the
 * current route or, if that's not available, the `taskId` prop. The
 * immediately available value in the redux store will be given first, but
 * a current copy of the task will also be fetched and passed down when
 * available.  A `completeTask` function is also made available to the wrapped
 * component, which can be used to mark the current task as complete with a
 * given status.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
const WithCurrentTask = WrappedComponent =>
  connect(mapStateToProps, mapDispatchToProps)(WithLoadedTask(WrappedComponent))

/**
 * WithLoadedTask is a private HOC used to fetch an up-to-date copy of the task
 * and parent challenge from the server.
 *
 * @private
 */
const WithLoadedTask = function(WrappedComponent) {
  return class extends Component {
    loadNeededTask = props => {
      if (_isFinite(props.taskId)) {
        props.loadTask(props.taskId)
      }
    }

    componentDidMount() {
      this.loadNeededTask(this.props)
    }

    componentWillReceiveProps(nextProps) {
      if (nextProps.taskId !== this.props.taskId) {
        this.loadNeededTask(nextProps)
      }
    }

    render() {
      // We don't need to pass anything down. WithCurrentTask grabs the latest
      // from the redux store, which is where our updated copy will end up.
      return <WrappedComponent {..._omit(this.props, 'loadTask')} />
    }
  }
}

export const mapStateToProps = (state, ownProps) => {
  const mappedProps = {task: null}

  const taskId = taskIdFromRoute(ownProps, ownProps.taskId)
  if (_isFinite(taskId)) {
    mappedProps.taskId = taskId
    const taskEntity = _get(state, `entities.tasks.${taskId}`)

    if (taskEntity) {
      // denormalize task so that parent challenge is embedded.
      mappedProps.task =
        denormalize(taskEntity, taskDenormalizationSchema(), state.entities)

      mappedProps.challengeId = _get(mappedProps.task, 'parent.id')
    }
  }

  return mappedProps
}

export const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    /**
     * For the LoadCurrentTask private HOC.
     *
     * @private
     */
    loadTask: _debounce(taskId => dispatch(loadCompleteTask(taskId)),
                        FRESHNESS_THRESHOLD),

    /**
     * Invoke to mark as a task as complete with the given status
     */
    completeTask: (taskId, challengeId, taskStatus, comment, taskLoadBy) => {
      dispatch(
        completeTask(taskId, challengeId, taskStatus)
      ).then(() => {
        if (_isString(comment) && comment.length > 0) {
          dispatch(addTaskComment(taskId, comment, taskStatus))
        }
        dispatch(fetchChallengeActions(challengeId))

        // If working on a virtual challenge, renew it (extend its expiration)
        // since we've seen some activity.
        if (_isFinite(ownProps.virtualChallengeId)) {
          dispatch(renewVirtualChallenge(ownProps.virtualChallengeId))
        }
      })

      // Load the next task from the challenge.
      nextRandomTask(dispatch, ownProps, taskId, taskLoadBy).then(newTask =>
        visitNewTask(ownProps, taskId, newTask)
      )
    },

    /**
     * Move to the next task without setting any completion status,
     * useful for when a user visits a task that is already complete.
     */
    nextTask: (challengeId, taskId, taskLoadBy, comment) => {
      if (_isString(comment) && comment.length > 0) {
        dispatch(addTaskComment(taskId, comment))
      }

      nextRandomTask(dispatch, ownProps, taskId, taskLoadBy).then(newTask =>
        visitNewTask(ownProps, taskId, newTask)
      )
    },

    contactTaskOwnerURL: ownerOSMId => contactOSMUserURL(ownerOSMId),
  }
}

/**
 * Retrieve the task id from the route, falling back to the given defaultId if
 * none is available.
 */
export const taskIdFromRoute = (props, defaultId) => {
  const taskId = parseInt(_get(props, 'match.params.taskId'), 10)
  return _isFinite(taskId) ? taskId : defaultId
}

/**
 * Retrieve the challenge id from the route, falling back to the given
 * defaultId if none is available.
 */
export const challengeIdFromRoute = (props, defaultId) => {
  const challengeId =
    parseInt(_get(props, 'match.params.challengeId'), 10)

  return _isFinite(challengeId) ? challengeId : defaultId
}

/**
 * Load a new random task, handling the differences between standard challenges
 * and virtual challenges.
 */
export const nextRandomTask = (dispatch, props, currentTaskId, taskLoadBy) => {
  // We need to make different requests depending on whether we're working on a
  // virtual challenge or a standard challenge.
  if (_isFinite(props.virtualChallengeId)) {
    return dispatch(
      loadRandomTaskFromVirtualChallenge(
        props.virtualChallengeId,
        taskLoadBy === TaskLoadMethod.proximity ? currentTaskId : undefined
      )
    )
  }
  else {
    return dispatch(
      loadRandomTaskFromChallenge(
        challengeIdFromRoute(props, props.challengeId),
        taskLoadBy === TaskLoadMethod.proximity ? currentTaskId : undefined
      )
    )
  }
}

/**
 * Route to the given new task, if valid. Otherwise route back to the home
 * page.
 */
export const visitNewTask = function(props, currentTaskId, newTask) {
  if (_isObject(newTask) && newTask.id !== currentTaskId) {
    // The route we use is different for virtual challenges vs standard
    // challenges.
    if (_isFinite(props.virtualChallengeId)) {
      props.history.push(`/virtual/${props.virtualChallengeId}/task/${newTask.id}`)
    }
    else {
      const challengeId = challengeIdFromRoute(props, props.challengeId)
      props.history.push(`/challenge/${challengeId}/task/${newTask.id}`)
    }
  }
  else {
    // Probably no tasks left in this challenge, back to challenges.
    props.history.push('/')
  }
}

export default WithCurrentTask
