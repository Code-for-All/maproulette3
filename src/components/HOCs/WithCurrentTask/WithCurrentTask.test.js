import React, { Component } from 'react'
import _each from 'lodash/each'
import { denormalize } from 'normalizr'
import { mapStateToProps, mapDispatchToProps, visitNewTask } from './WithCurrentTask'
import { taskDenormalizationSchema,
         loadCompleteTask,
         loadRandomTaskFromChallenge,
         addTaskComment,
         completeTask } from '../../../services/Task/Task'
import { fetchChallengeActions } from '../../../services/Challenge/Challenge'

jest.mock('normalizr')
jest.mock('../../../services/Task/Task')
jest.mock('../../../services/Challenge/Challenge')

denormalize.mockImplementation((challenge, schema, entities) => challenge)

let basicState = null

beforeEach(() => {
  basicState = {
    entities: {
      tasks: {
        1: {
          taskId: 1,
          name: "task1",
          parent: {
            id: 123
          }
        }
      },
      challenges: [
        {
          id: 123,
          name: "challenge1",
          difficulty: "hard",
          actions: {available: 3},
          enabled: true,
        },
      ]
    },
    currentPreferences: {
      challenges: {
        123: {
          minimize: true,
          collapseInstructions: true,
        }
      }
    },
  }
})

test("mapStateToProps maps task from a given taskId", () => {
  const mappedProps = mapStateToProps(basicState, {taskId: 1})

  // { task: { taskId: 1, name: 'task1', parent: { id: 5 } },
  //       taskId: 1,
  //       minimizeChallenge: false,
  //       collapseInstructions: false }

  expect(mappedProps.task).toEqual(basicState.entities.tasks["1"])
  expect(denormalize).toHaveBeenCalled()

  expect(mappedProps).toMatchSnapshot()
})


test("mapStateToProps maps minimizeChallenge to current minimize preference", () => {
  basicState.currentPreferences.challenges["123"].minimize = false
  const mappedProps = mapStateToProps(basicState, {taskId: 1})

  expect(mappedProps.minimizeChallenge).toEqual(false)
})

test("mapStateToProps maps collapseInstructions to current minimize preference", () => {
  basicState.currentPreferences.challenges["123"].collapseInstructions = false
  const mappedProps = mapStateToProps(basicState, {taskId: 1})

  expect(mappedProps.collapseInstructions).toEqual(false)
})

test("mapDispatchToProps maps some functions", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {})

  expect(mappedProps).toMatchSnapshot()
})

test("mapDispatchToProps loadTask calls loadCompleteTask", () => {
  const dispatch  = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {})

  mappedProps.loadTask(1)
  expect(dispatch).toBeCalled()
  expect(loadCompleteTask).toBeCalledWith(1)
})

test("mapDispatchToProps completeTask calls completeTask", () => {
  const task = {id: 1}
  const dispatch  = jest.fn(() => Promise.resolve())
  const history = {
   push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  mappedProps.completeTask(1, 123, "good", "")
  expect(dispatch).toBeCalled()
  expect(completeTask).toBeCalledWith(1, 123, "good")
})

test("completeTask calls addComment if comment present",  async () => {
  const dispatch  = jest.fn(() => Promise.resolve())
  const history = {
   push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  await mappedProps.completeTask(1, 123, "good", "my new comment")
  expect(addTaskComment).toHaveBeenLastCalledWith(1, "my new comment", "good")
})

test("completeTask does not call addComment if no comment", async () => {
  addTaskComment.mockClear()

  const dispatch  = jest.fn(() => Promise.resolve())
  const history = {
   push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  await mappedProps.completeTask(1, 123, "good", "")
  expect(addTaskComment).not.toHaveBeenCalled()
})

test("completeTask calls loadRandomTaskFromChallenge", () => {
  const dispatch  = jest.fn(() => Promise.resolve())
  const history = {
   push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  mappedProps.completeTask(1, 123, "good", "my comment")
  expect(loadRandomTaskFromChallenge).toBeCalledWith(123, 1)
})

test("completeTask calls fetchChallengeActions", () => {
  const taskId = 1
  const challengeId = 123

  const dispatch  = jest.fn(() => Promise.resolve())
  const history = {
   push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  mappedProps.completeTask(taskId, challengeId, "good", "my comment")
  expect(fetchChallengeActions).toBeCalledWith(challengeId)
})

test("completeTask routes the user to the new task if there is one", async () => {
  const taskId = 1
  const challengeId = 123
  const nextTask = {id: 2}

  completeTask.mockReturnValueOnce(Promise.resolve())
  loadRandomTaskFromChallenge.mockReturnValueOnce(Promise.resolve(nextTask))
  const dispatch = jest.fn(value => value)
  const history = {
    push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  await mappedProps.completeTask(taskId, challengeId, "good", "my comment")
  expect(history.push).toBeCalledWith(`/challenge/${challengeId}/task/${nextTask.id}`)
})

test("completeTask routes the user home if the next task isn't new", async () => {
  const taskId = 1
  const challengeId = 123
  const nextTask = {id: taskId} // same as our current task

  completeTask.mockReturnValueOnce(Promise.resolve())
  loadRandomTaskFromChallenge.mockReturnValueOnce(Promise.resolve(nextTask))
  const dispatch = jest.fn(value => value)
  const history = {
    push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  await mappedProps.completeTask(taskId, challengeId, "good", "my comment")
  expect(history.push).toBeCalledWith('/')
})

test("completeTask routes the user home if there is no next task", async () => {
  const taskId = 1
  const challengeId = 123

  completeTask.mockReturnValueOnce(Promise.resolve())
  loadRandomTaskFromChallenge.mockReturnValueOnce(Promise.resolve(null))
  const dispatch = jest.fn(value => value)
  const history = {
    push: jest.fn(),
  }

  const mappedProps = mapDispatchToProps(dispatch, {history})

  await mappedProps.completeTask(taskId, challengeId, "good", "my comment")
  expect(history.push).toBeCalledWith('/')
})
