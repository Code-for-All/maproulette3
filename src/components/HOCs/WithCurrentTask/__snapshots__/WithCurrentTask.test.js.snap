// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps maps some functions 1`] = `
Object {
  "completeTask": [Function],
  "loadTask": [Function],
  "nextTask": [Function],
}
`;

exports[`mapStateToProps maps task from the taskId in the route 1`] = `
Object {
  "challengeId": undefined,
  "task": Object {
    "id": 987,
    "name": "task987",
    "parent": 123,
  },
  "taskId": 987,
  "virtualChallengeId": undefined,
}
`;
